programs/stake-pool/src/instructions/initialize.rs

```diff
@@ -0,0 +1,199 @@
+};
+
+use crate::{state::Fee, write_bytes, UNINIT_BYTE};
+
+///   Initializes a new `StakePool`.
+///
+/// ### Accounts:
+///   0. `[WRITE]` New `StakePool` to create.
+///   1. `[SIGNER]` Manager
+///   2. `[]` Staker
+///   3. `[]` Stake pool withdraw authority
+///   4. `[WRITE]` Uninitialized validator stake list storage account
+///   5. `[]` Reserve stake account must be initialized, have zero balance,
+///      and staker / withdrawer authority set to pool withdraw authority.
+///   6. `[]` Pool token mint. Must have zero supply, owned by withdraw
+///      authority.
+///   7. `[]` Pool account to deposit the generated fee for manager.
+///   8. `[]` Token program id
+///   9. `[]` (Optional) Deposit authority that must sign all deposits.
+///      Defaults to the program address generated using
+///      `find_deposit_authority_program_address`, making deposits
+///      permissionless.
+pub struct Initialize<'a> {
+    /// Accounts
+    /// Stake Pool Account.
+    pub stake_pool: &'a AccountInfo,
+    /// Manager Account.
+    pub manager: &'a AccountInfo,
+    /// Staker Account.
+    pub staker: &'a AccountInfo,
+    /// Withdraw Authority Account.
+    pub stake_pool_withdraw_authority: &'a AccountInfo,
+    /// Validator list Account.
+    pub validator_list: &'a AccountInfo,
+    /// Reserve stake Account.
+    pub reserve_stake: &'a AccountInfo,
+    /// Pool mint Account.
+    pub pool_mint: &'a AccountInfo,
+    /// Manager Pool Account.
+    pub manager_pool_account: &'a AccountInfo,
+    /// Token program.
+    pub token_program: &'a AccountInfo,
+    /// Deposit Authority (Optional)
+    pub deposit_authority: Option<&'a AccountInfo>,
+
+    /// input
+    /// Amount.
+    pub amount: u64,
+    /// Deposit Authority Pubkey(Optional)
+    pub deposit_authority_key: Option<&'a Pubkey>,
+    /// Fee.
+    pub fee: Fee,
+    /// Withdrawal Fee.
+    pub withdrawal_fee: Fee,
+    /// Deposit Fee.
+    pub deposit_fee: Fee,
+    /// Referral Fee.
+    pub referral_fee: u8,
+    /// Max validators.
+    pub max_validators: u32,
+}
+
+/// Seed for deposit authority seed
+const AUTHORITY_DEPOSIT: &[u8] = b"deposit";
+
+/// Generates the deposit authority program address for the stake pool
+pub fn find_deposit_authority_program_address(
+    program_id: &Pubkey,
+    stake_pool_address: &Pubkey,
+) -> (Pubkey, u8) {
+    find_program_address(
+        &[stake_pool_address.as_ref(), AUTHORITY_DEPOSIT],
+        program_id,
+    )
+}
+
+impl Initialize<'_> {
+    #[inline(always)]
+    pub fn invoke(&self) -> ProgramResult {
+        self.invoke_signed(&[])
+    }
+
+    pub fn invoke_signed(&self, signers: &[Signer]) -> ProgramResult {
+        let deposit_authority: Pubkey;
+
+        if let Some(info) = self.deposit_authority {
+            deposit_authority = *info.key();
+        } else {
+            deposit_authority =
+                find_deposit_authority_program_address(&crate::ID, &self.stake_pool.key()).0;
```

Comment:
If the `deposit_authority` has not being informed, then no need to set one nor derive its address.